src/Cool/CollaboraConnection.php

```diff
@@ -0,0 +1,132 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * Copyright the Collabora Online contributors.
+ *
+ * SPDX-License-Identifier: MPL-2.0
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+namespace Drupal\collabora_online\Cool;
+
+use Drupal\collabora_online\Exception\CollaboraNotAvailableException;
+use Drupal\Core\Config\ConfigFactoryInterface;
+use Drupal\Core\Logger\LoggerChannelInterface;
+use GuzzleHttp\ClientInterface;
+use GuzzleHttp\RequestOptions;
+use Psr\Http\Client\ClientExceptionInterface;
+use Symfony\Component\DependencyInjection\Attribute\Autowire;
+use Symfony\Component\HttpFoundation\RequestStack;
+
+/**
+ * Service to load the discovery.xml from the Collabora server.
+ */
+class CollaboraConnection {
+
+  /**
+   * Constructor.
+   *
+   * @param \Drupal\Core\Logger\LoggerChannelInterface $logger
+   *   Logger channel.
+   * @param \Drupal\Core\Config\ConfigFactoryInterface $configFactory
+   *   Config factory.
+   * @param \Symfony\Component\HttpFoundation\RequestStack $requestStack
+   *   Request stack.
+   * @param \GuzzleHttp\ClientInterface $client
+   *   Http client.
+   */
+  public function __construct(
+    #[Autowire(service: 'logger.channel.collabora_online')]
+    protected readonly LoggerChannelInterface $logger,
+    protected readonly ConfigFactoryInterface $configFactory,
+    protected readonly RequestStack $requestStack,
+    protected readonly ClientInterface $client,
+  ) {}
+
+  /**
+   * Loads the WOPI server url from configuration.
+   *
+   * @return string
+   *   Base URL to access the WOPI server from Drupal.
+   *
+   * @throws \Drupal\collabora_online\Exception\CollaboraNotAvailableException
+   *   The WOPI server url is misconfigured, or the protocol does not match
+   *   that of the current Drupal request.
+   */
+  protected function getWopiClientServerBaseUrl(): string {
+    $cool_settings = $this->loadSettings();
+    $wopi_client_server = $cool_settings['server'] ?? NULL;
+    if (!$wopi_client_server) {
+      throw new CollaboraNotAvailableException('The configured Collabora Online server address is empty.');
+    }
+    $wopi_client_server = trim($wopi_client_server);
+
+    if (!preg_match('@^https?://@', $wopi_client_server)) {
+      throw new CollaboraNotAvailableException(sprintf(
+        "The configured Collabora Online server address must begin with 'http://' or 'https://'. Found '%s'.",
+        $wopi_client_server,
+      ));
+    }
```

Comment:
All this validation of config values seems a bit out of place.
In most of Drupal core, config values are simply trusted - for better or worse.

But, currently the config form itself does not really validate these settings.
We do have a schema, but in the config form I can just enter any value I like for the server and host url, e.g. "abc".

Imo we should keep the validation as-is for now, and only change it once we have proper validation in the form.

(The validation is coming from the original code, it has been moved around and refactored but it is still mostly equivalent with the original code.)